# Generated by using Rcpp::compileAttributes() -> do not edit by hand
# Generator token: 10BE3573-1514-4C36-9D1C-5A225CD40393

#' @title anccall_c: Generating ancestry call along the chromosome
#'
#' @param deltaf The cut-off for minimum delta f between two reference populations
#' @param window Size of the sliding window
#' @param SMAX The number of all snp sites
#' @param anclikdir the dir to anclik file
#' @param output the dir of the output ancfreq file
anccall_c <- function(deltaf, window, SMAX, anclikdir, output, chrom, indiv, mode, n) {
    invisible(.Call(`_rLCLAE_anccall_c`, deltaf, window, SMAX, anclikdir, output, chrom, indiv, mode, n))
}

#' @title ancfreq_c
#' @description Calculating the allele frequency in two different ancestral population
#'
#' @param n an integer, the number of individuals contain in the input
#' @param type 1 or 2, specifying diploid/haploid data.
#' @param pop1 the dir of a file containing individuals belonging to reference population 1
#' @param pop2 the dir of a file containing individuals belonging to reference population 2
#' @param input the dir of the input genolik file
#' @param output the dir of the output ancfreq file
ancfreq_c <- function(n, type, pop1, pop2, input, output) {
    invisible(.Call(`_rLCLAE_ancfreq_c`, n, type, pop1, pop2, input, output))
}

#' ancfreq_merge: Merging haploid & dipoid dataset
#'
#' @param hapfreq The dir to file containing ancfreq for haploid data
#' @param dipfreq The dir to file containing ancfreq for diploid data
#' @param type An integer, 0-intersection 1-keep all haploid snp 2-keep all diploid snp 3-union
#' @param outputdir The output dir
ancfreq_merge <- function(hapfreq, dipfreq, outputdir, type) {
    invisible(.Call(`_rLCLAE_ancfreq_merge`, hapfreq, dipfreq, outputdir, type))
}

#' anclik_c: Calculating the ancestral likelihood of a test individual
#'
#' @param sum an integer, the number of individuals contain in the input
#' @param type 1 or 2, specifying diploid/haploid data.
#' @param testid the individual to be estimate
#' @param genolik the dir to genolik file
#' @param ancfreq the dir to ancfreq file
#' @param output the dir of the output ancfreq file
anclik_c <- function(sum, type, testid, genolik, ancfreq, output) {
    invisible(.Call(`_rLCLAE_anclik_c`, sum, type, testid, genolik, ancfreq, output))
}

#' @title anctract_c: Generating ancestry tract from call.
#'
#' @param chrlen Length of the chromosome
#' @param excludelen length to be excluded
#' @param input input file dir
#' @param output output file dir
anctract_c <- function(chrlen, excludelen, input, output) {
    invisible(.Call(`_rLCLAE_anctract_c`, chrlen, excludelen, input, output))
}

#' filt1_dip: transforming the phred score to relative genotype likelihood
#'
#' @param n an integer, the number of individuals contain in the vcf
#' @param skip an integer, lines to skip
#' @param dur an ingerger, lines for current chromosome
#' @param input a string, the dir to cleaned vcf file
#' @param output a string, the dir to the output file
filt1_dip <- function(n, skip, dur, input, output) {
    invisible(.Call(`_rLCLAE_filt1_dip`, n, skip, dur, input, output))
}

#' filt1_hap: transforming the phred score to relative genotype likelihood
#'
#' @param n an integer, the number of individuals contain in the vcf
#' @param skip an integer, lines to skip
#' @param dur an ingerger, lines for current chromosome
#' @param input a string, the dir to cleaned vcf file
#' @param output a string, the dir to the output file
filt1_hap <- function(n, skip, dur, input, output) {
    invisible(.Call(`_rLCLAE_filt1_hap`, n, skip, dur, input, output))
}

#' GL=10^(-pl)
#'
#' @param x A single double.
glpow <- function(x) {
    .Call(`_rLCLAE_glpow`, x)
}

